foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/asm/ASMFactory.java

```diff
@@ -0,0 +1,361 @@
+        } else {
+            throw ValidationException.notAvailableASMService();
+        }
+    }
+
+    public static Label createLabel() {
+        String asmService = ASMFactory.getAsmService();
+        if (ASM_SERVICE_ECLIPSELINK.equals(asmService)) {
+            return new org.eclipse.persistence.asm.internal.platform.eclipselink.LabelImpl();
+        } else if (ASM_SERVICE_OW2.equals(asmService)) {
+            return new org.eclipse.persistence.asm.internal.platform.ow2.LabelImpl();
+        } else {
+            throw ValidationException.notAvailableASMService();
+        }
+    }
+
+    public static SerialVersionUIDAdder createSerialVersionUIDAdder(final ClassVisitor classVisitor) {
+        String asmService = ASMFactory.getAsmService();
+        if (ASM_SERVICE_ECLIPSELINK.equals(asmService)) {
+            return new org.eclipse.persistence.asm.internal.platform.eclipselink.SerialVersionUIDAdderImpl(classVisitor);
+        } else if (ASM_SERVICE_OW2.equals(asmService)) {
+            return new org.eclipse.persistence.asm.internal.platform.ow2.SerialVersionUIDAdderImpl(classVisitor);
+        } else {
+            throw ValidationException.notAvailableASMService();
+        }
+    }
+
+    public static String getAsmService() {
+        String asmService = PrivilegedAccessHelper.getSystemProperty(SystemProperties.ASM_SERVICE);
+        if (asmService != null) {
+            if (ASM_SERVICE_ECLIPSELINK.equals(asmService) && isASMImplementationAvailable(ASM_ECLIPSELINK_CLASS_VISITOR)) {
+                LOG.finest("EclipseLink ASM implementation is used.");
+                return asmService;
+            } else if (ASM_SERVICE_OW2.equals(asmService) && isASMImplementationAvailable(ASM_OW2_CLASS_VISITOR)) {
+                LOG.finest("OW2 ASM implementation is used.");
+                return asmService;
+            } else {
+                throw ValidationException.incorrectASMServiceProvided();
+            }
+        }
+        //Fallback to default if ASM service is not specified
+        if (isASMImplementationAvailable(ASM_ECLIPSELINK_CLASS_VISITOR)) {
+            LOG.finest("EclipseLink ASM implementation is used.");
+            return ASM_SERVICE_ECLIPSELINK;
+        } else if (isASMImplementationAvailable(ASM_OW2_CLASS_VISITOR)) {
+            LOG.finest("OW2 ASM implementation is used.");
+            return ASM_SERVICE_OW2;
+        } else {
+            throw ValidationException.notAvailableASMService();
+        }
+    }
+
+    private static boolean isASMImplementationAvailable(String className) {
+        try {
+            PrivilegedAccessHelper.getClassForName(className);
+            return true;
+        } catch (ClassNotFoundException e) {
+            return false;
+        }
+    }
+
+    static int getLatestOPCodeVersion() {
+        final Map<String, Integer> versionMap = new LinkedHashMap<>();
+        Pattern searchPattern = Pattern.compile("^V\\d((_\\d)?|\\d*)");
```

Comment:
Can searchPattern be compiled as a static outside of this method?  That'd be more performant.